File: src/main/java/AgentProcessor.java
public class AgentProcessor {
    public Agent process(final Agent agent, final double brandFactor) {
        final Agent.AgentBuilder builder = new Agent.AgentBuilder().withAgent(agent);

        builder.withAge(agent.getAge() + 1);

        if (agent.getAutoRenew() == 0) {
            double random = Math.random() * 3;
            double affinity = agent.getPaymentAtPurchase() / agent.getAttributePrice()
                    + random * agent.getAttributePromotion() * agent.getInertiaForSwitch();
            if (Agent.BreedType.BREED_C == agent.getAgentBreed()
                    && affinity < (agent.getSocialGrade() * agent.getAttributeBrand())) {
                builder.withAgentBreed(Agent.BreedType.BREED_NC);
            } else if (Agent.BreedType.BREED_NC == agent.getAgentBreed()
                    && affinity < (agent.getSocialGrade() * agent.getAttributeBrand() * brandFactor)) {
                builder.withAgentBreed(Agent.BreedType.BREED_C);
            }
        }
        return builder.addPreviousAgent(agent).build();
    }
}

File: src/main/scala/com/simudyne/platform/AgentProcessor.scala
package com.simudyne.platform

trait AgentProcessor {
  def process(agent: Agent, brandFactor: Double): Agent = {
    if (agent.autoRenew == 1) {
      val random = Math.random() * 3
      val affinity = agent.paymentAtPurchase / agent.attributePrice + random * agent.attributePromotion * agent.inertiaForSwitch
      if (BreedType.BREED_C == agent.agentBreed
        && affinity < (agent.socialGrade * agent.attributeBrand)) {
        Agent(agent, BreedType.BREED_NC)
      } else if (BreedType.BREED_NC == agent.agentBreed && affinity < (agent.socialGrade * agent.attributeBrand * brandFactor)) {
        Agent(agent, BreedType.BREED_C)
      } else {
        Agent(agent)
      }
    } else {
      Agent(agent)
    }
  }
}

object AgentProcessor extends AgentProcessor